crates/kubelet/src/state.rs

```diff
@@ -104,3 +291,61 @@ impl<P: 'static + Sync + Send> State<P> for Stub {
         Ok(serde_json::json!(null))
     }
 }
+
+#[cfg(test)]
+mod test {
+    use crate::pod::Pod;
+    use crate::state::{EdgeTo, State, Transition};
+
+    #[derive(Debug)]
+    struct PodState;
+
+    #[derive(Debug)]
+    struct ValidState;
+
+    #[async_trait::async_trait]
+    impl State<PodState> for ValidState {
+        async fn next(
+            &self,
+            _pod_state: &mut PodState,
+            _pod: &Pod,
+        ) -> anyhow::Result<Transition<PodState>> {
+            Ok(Transition::Complete(Ok(())))
+        }
+
+        async fn json_status(
+            &self,
+            _pod_state: &mut PodState,
+            _pod: &Pod,
+        ) -> anyhow::Result<serde_json::Value> {
+            Ok(serde_json::json!(null))
+        }
+    }
+
+    #[test]
```

Comment:
I think I'm missing the test here.  It seems to verify that the `impl for TestState` compiles, but I'm missing where it checks that a pod operating under this state machine moves correctly through the states.  (This is quite possibly me missing things.)

More significantly, I can't think of a way to test that the `EdgeTo` mechanism is serving its purpose by _preventing_ the state machine from traversing undeclared edges.  This would involve asserting that a piece of code does _not_ compile and I'm not sure how to do that short of invoking the Rust compiler from the test itself e.g. via a process exec call.

Reply:
So, the real tests are up in the doctest for for `Transition`, where we are actually able to flag a test as `compile_fail` and confirm that the compiler is checking these things. This test mostly serves as a sanity check that the `compile_fail` isnt due to something else (change in API). I think this test is mostly superfluous, as `Stub`, as well as some of the doctests test roughly the same thing. I just kept it around because we may want to expand this test module in the future and it didnt hurt.